Google.Apis.Common/src/Google.Apis.Common/PagingAsyncEnumerable.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright 2015 Google Inc. All Rights Reserved.
+// Licensed under the Apache License Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Common
+{
+    /// <summary>
+    /// Implementation of <see cref="IAsyncEnumerable{T}"/> used for pagination.
+    /// </summary>
+    internal sealed class PagingAsyncEnumerable<T> : IAsyncEnumerable<T>
+    {
+        private readonly Func<CancellationToken, Task<IEnumerable<T>>> provider;
+
+        internal PagingAsyncEnumerable(Func<CancellationToken, Task<IEnumerable<T>>> provider)
+        {
+            this.provider = provider;
+        }
+
+        public IAsyncEnumerator<T> GetEnumerator()
+        {
+            return new PagingAsyncEnumerator<T>(provider);
+        }
+    }
+
+    internal sealed class PagingAsyncEnumerator<T> : IAsyncEnumerator<T>
+    {
+        private static readonly List<T> emptyList = new List<T>();
+
+        private readonly Func<CancellationToken, Task<IEnumerable<T>>> provider;
+
+        /// <summary>
+        /// The list of items we're currently working our way through, before requesting more.
+        /// If this is a null refernece, we have exhausted the source.
+        /// </summary>
+        private List<T> items = emptyList;
+        /// <summary>
+        /// The current index in the list. If this is equal to the length of the list, we need
+        /// to fetch more items before returning any.
+        /// </summary>
+        private int currentIndex;
```

Comment:
I wonder if we could re-use the IEnumerator<T> implementation of the List<T> here instead of using the index. Then invoke MoveNext on that instead. Seems a bit more direct implementation.